baseband/guppi/base.py

```diff
@@ -0,0 +1,360 @@
+                                        header=header)
+        return GUPPIFrame(header, payload)
+
+
+class GUPPIStreamBase(VLBIStreamBase):
+    """Base for GUPPI streams."""
+
+    _sample_shape_maker = GUPPIPayload._sample_shape_maker
+
+    def __init__(self, fh_raw, header0, squeeze=True, subset=()):
+
+        # GUPPI sample positions are determined by "packets", the bytesize of
+        # which are given in the header.
+        self._packets_per_frame = (
+            (header0.payload_nbytes - header0.overlap * header0._bpcs // 8) //
+            header0['PKTSIZE'])
+
+        # Set samples per frame to valid ones only.
+        samples_per_frame = header0.samples_per_frame - header0.overlap
+
+        super(GUPPIStreamBase, self).__init__(
+            fh_raw=fh_raw, header0=header0, sample_rate=header0.sample_rate,
+            samples_per_frame=samples_per_frame,
+            unsliced_shape=header0.sample_shape, bps=header0.bps,
+            complex_data=header0.complex_data, squeeze=squeeze, subset=subset,
+            fill_value=0.)
+
+    # Overriding so the docstring indicates the exclusion of the overlap.
+    samples_per_frame = property(VLBIStreamBase.samples_per_frame.fget,
+                                 VLBIStreamBase.samples_per_frame.fset,
+                                 doc=("Number of complete samples per frame, "
+                                      "excluding overlap."))
+
+
+class GUPPIStreamReader(GUPPIStreamBase, VLBIStreamReaderBase):
+    """DADA format reader.
+
+    Allows access to DADA files as a continuous series of samples.
+
+    Parameters
+    ----------
+    fh_raw : filehandle
+        Filehandle of the raw DADA stream.
+    squeeze : bool, optional
+        If `True` (default), remove any dimensions of length unity from
+        decoded data.
+    subset : indexing object or tuple of objects, optional
+        Specific components of the complete sample to decode (after possibly
+        squeezing).  If a single indexing object is passed, it selects
+        polarizations.  With a tuple, the first selects polarizations and the
+        second selects channels.  If the tuple is empty (default), all
+        components are read.
+    """
+    def __init__(self, fh_raw, squeeze=True, subset=()):
+        fh_raw = GUPPIFileReader(fh_raw)
+        header0 = GUPPIHeader.fromfile(fh_raw)
+        super(GUPPIStreamReader, self).__init__(fh_raw, header0,
+                                                squeeze=squeeze, subset=subset)
+
+    @lazyproperty
+    def _last_header(self):
+        """Header of the last file for this stream."""
+        nframes, fframe = divmod(self.fh_raw.seek(0, 2),
+                                 self.header0.frame_nbytes)
+        # If there is a non-integer number of frames, assume it's the last
+        # frame that's missing bytes, and go to the last full frame.
+        if fframe:
+            self.fh_raw.seek((nframes - 1) * self.header0.frame_nbytes)
```

Comment:
I had thought that seeking forward would be less accurate than seeking from the end for very large files.   Is that an irrational fear since ints in Python can be arbitrarily large without losing accuracy?

Reply:
Internally, I'm sure it just does a seek from the start with (size + offset) - i.e., I wouldn't worry.